File: apple/Sources/SignetAppleCore/SecurityTier.swift
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: 2026 Nirapod Labs

/// The hardware backing a signing key, reported as the achieved level and never
/// assumed from the request. Closed set; see `conformance/security-level.json`.
public enum SecurityLevel: String, Sendable, Equatable, CaseIterable {
    case secureEnclave
    case strongBox
    case tee
    case tpm
    case software
}

/// How the achieved `SecurityLevel` was determined. Only `attested` is
/// cryptographic proof; every other value is an on-device self-report.
public enum TierEvidence: String, Sendable, Equatable, CaseIterable {
    case attested
    case keyInfoReadback
    case seTokenPresent
    case inferred
    case selfReportUnverified
}

/// The presence check bound to a key at creation, derived from the created
/// access control and never echoed from the request.
public enum AuthClass: String, Sendable, Equatable, CaseIterable {
    case none
    case biometricOnly
    case biometricOrDeviceCredential
    case deviceCredentialOnly
}

/// A class in the tier partial order. `atLeast(_:)` selects by class:
/// `discreteSecure` covers Secure Enclave, StrongBox, and TPM, and outranks
/// `trustedEnvironment`.
public enum HardwareClass: String, Sendable, Equatable, CaseIterable {
    case discreteSecure
    case trustedEnvironment
}

/// Tier selection on `KeySpec`. Selection is by class, never a concrete
/// `SecurityLevel`; the achieved level is reported in `SecurityTierReport`.
public enum TierPolicy: Sendable, Equatable {
    /// The device's best hardware tier. Fails `unavailableTier` if no hardware
    /// is available; never returns software.
    case strongest
    /// A hard floor by class. Fails `unavailableTier` below the class.
    case atLeast(HardwareClass)
    /// Never fails on tier where a software keystore exists; may then return a
    /// weaker level with `meetsFloor == false` and honest evidence (the only
    /// policy that can return software). On a platform with no software backend
    /// (Apple), yields the Secure Enclave or fails `unavailableTier`.
    case bestEffort
}

extension HardwareClass {
    /// Rank in the partial order; lower is stronger.
    var rank: Int {
        switch self {
        case .discreteSecure: return 0
        case .trustedEnvironment: return 1
        }
    }
}

extension SecurityLevel {
    /// The partial-order class this level belongs to, or `nil` for `software`,
    /// which is below every hardware class.
    var hardwareClass: HardwareClass? {
        switch self {
        case .secureEnclave, .strongBox, .tpm: return .discreteSecure
        case .tee: return .trustedEnvironment
        case .software: return nil
        }
    }
}

/// One report shape everywhere. `achieved` is read back from the created key,
/// `meetsFloor` derives from the tier partial order, and `authEnforced` is
/// derived from the created access control.
public struct SecurityTierReport: Sendable, Equatable {
    public let achieved: SecurityLevel
    public let requested: TierPolicy
    public let meetsFloor: Bool
    public let evidence: TierEvidence
    public let authEnforced: AuthClass
    public let invalidated: Bool
    public let schemaVersion: Int

    public init(
        achieved: SecurityLevel,
        requested: TierPolicy,
        meetsFloor: Bool,
        evidence: TierEvidence,
        authEnforced: AuthClass,
        invalidated: Bool,
        schemaVersion: Int = 1
    ) {
        self.achieved = achieved
        self.requested = requested
        self.meetsFloor = meetsFloor
        self.evidence = evidence
        self.authEnforced = authEnforced
        self.invalidated = invalidated
        self.schemaVersion = schemaVersion
    }
}

extension TierPolicy {
    /// Whether `achieved` satisfies this policy's floor, per the partial order.
    func isMet(by achieved: SecurityLevel, platformStrongest: SecurityLevel) -> Bool {
        switch self {
        case .strongest:
            return achieved == platformStrongest
        case .atLeast(let floor):
            guard let achievedClass = achieved.hardwareClass else { return false }
            return achievedClass.rank <= floor.rank
        case .bestEffort:
            return achieved.hardwareClass == .discreteSecure
        }
    }
}

File: android/src/main/kotlin/org/nirapod/signet/SecurityTier.kt
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: 2026 Nirapod Labs

package org.nirapod.signet

/**
 * The hardware backing a signing key, reported as the achieved level and never
 * assumed from the request. Closed set; see `conformance/security-level.json`.
 */
public enum class SecurityLevel {
    strongBox,
    tee,
}

/**
 * How the achieved [SecurityLevel] was determined. The Android core reads the
 * level back from the created key via the platform Keystore.
 */
public enum class TierEvidence {
    keyInfoReadback,
}

/**
 * The presence check bound to a key at creation, derived from the created key
 * and never echoed from the request.
 */
public enum class AuthClass {
    none,
    biometricOnly,
    biometricOrDeviceCredential,
    deviceCredentialOnly,
}

/**
 * A class in the tier partial order. `atLeast` selects by class: `discreteSecure`
 * covers StrongBox and outranks `trustedEnvironment`, which covers the TEE.
 */
public enum class HardwareClass(internal val rank: Int) {
    discreteSecure(0),
    trustedEnvironment(1),
}

/**
 * The partial-order class a level belongs to. Total over the hardware-only level
 * set: StrongBox is `discreteSecure`, the TEE is `trustedEnvironment`.
 */
internal val SecurityLevel.hardwareClass: HardwareClass
    get() = when (this) {
        SecurityLevel.strongBox -> HardwareClass.discreteSecure
        SecurityLevel.tee -> HardwareClass.trustedEnvironment
    }

/**
 * Tier selection on [KeySpec]. Selection is by class, never a concrete
 * [SecurityLevel]; the achieved level is reported in [SecurityTierReport].
 */
public sealed class TierPolicy {
    /**
     * The device's best hardware tier. A StrongBox device that falls back to the
     * TEE fails `unavailableTier` rather than silently downgrade.
     */
    public object Strongest : TierPolicy()

    /** A hard floor by class. Fails `unavailableTier` below the class. */
    public data class AtLeast(val hardwareClass: HardwareClass) : TierPolicy()

    /** Whether [achieved] satisfies this policy's floor, per the partial order. */
    internal fun isMet(achieved: SecurityLevel, platformStrongest: SecurityLevel): Boolean =
        when (this) {
            Strongest -> achieved == platformStrongest
            is AtLeast -> achieved.hardwareClass.rank <= hardwareClass.rank
        }
}

/**
 * One report shape everywhere. `achieved` is read back from the created key and
 * `authEnforced` is derived from the created key.
 *
 * `requested` is optional: `generateKey` populates it; a `getSecurityTier`
 * re-read leaves it null, since the policy is not stored with the key. The
 * Android core reads `authEnforced` back from the key. A null `authEnforced` in
 * the shared report shape means unobservable, distinct from [AuthClass.none] (a
 * key created with no presence check).
 */
public data class SecurityTierReport(
    val achieved: SecurityLevel,
    val requested: TierPolicy?,
    val evidence: TierEvidence,
    val authEnforced: AuthClass?,
    val invalidated: Boolean,
    val schemaVersion: Int = 1,
)
